feat: aztec-up can be used to manage multiple aztec versions

charlielye opened the pull request:
`aztec-up` can be used to manage multiple versions.

```
$ aztec-up
aztec-up - Aztec version manager

Usage: aztec-up [command] [options]

Commands:
  install <version>         Install a version and switch to it
  use [<version>]           Switch to an installed version (or read from .aztecrc)
  list                      List installed versions
  uninstall <version>       Remove an installed version
  self-update               Update aztec-up itself to the latest version

Options:
  -h, --help                Show this help message

Examples:
  aztec-up install          Install the latest version
  aztec-up install nightly  Install the latest nightly version
  aztec-up install 0.85.0   Install a specific version
  aztec-up use 0.85.0       Switch to version 0.85.0
  aztec-up use              Read version from .aztecrc and switch to it
  aztec-up list             Show all installed versions
  aztec-up self-update      Update aztec-up to latest
```

If a `.aztecrc` file is present in the cwd, `aztec` will automatically switch to that version in that run context.

```
t % aztec --version
0.0.1-commit.1142ef1
t % cd fooproj
t % cat .aztecrc
0.0.1-commit.6d3c34e
t % aztec --version
0.0.1-commit.6d3c34e
t %
```

# Testing

You can install `aztec-up` with:
```
VERSION=0.0.1-commit.6d3c34e bash -i <(curl -s https://install.aztec.network/0.0.1-commit.6d3c34e/aztec-install)
```

At present you can't list available remote versions. The code does currently support listing available remote aliases, but they're not deployed yet. You can install another version with e.g.
```
aztec-up install 0.0.1-commit.1142ef1
```

Then list versions with:

```
t % aztec-up list
Installed versions:

* 0.0.1-commit.1142ef1 (current)
  0.0.1-commit.6d3c34e

Aliases:
  (could not fetch aliases)
```

charlielye commented:
* **#19261** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19261?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19261?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#18665** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18665?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

olehmisar commented:
a suggestion:

does this make `aztec` command read from `.aztecrc` as well?

E.g., the ideal workflow (similar to [gaztec](https://github.com/nemi-fi/gaztec)):
```sh
bash -i <(curl -s https://install.aztec.network)
aztec --version
#> 2.1.9 # the latest version by default

echo 3.0.0 > .aztecrc
aztec --version
#> 3.0.0

# in another project
echo 3.0.0-devnet.5 > .aztecrc
aztec --version
#> 3.0.0-devnet.5
```

Note that `aztec-up` was never used. Calling `aztec` should automatically install the version from `.aztecrc` or fallback to the global installed version (when aztec-up was last called).

This mimics how [`rust-toolchain.toml`](https://rust-lang.github.io/rustup/overrides.html#the-toolchain-file) works. Maybe worth using `aztec-toolchain.toml` instead of `.aztecrc` to make it extendable in the future?

ludamad commented on ci3/dashboard/rk.py:
```diff
@@ -63,7 +64,7 @@ def read_breakdown_from_disk(runtime, flow_name, sha):
 @auth.verify_password
 def verify_password(username, password):
-    if username == "aztec" and password == "letmeseethoselogs":
+    if username == "aztec" and password == DASHBOARD_PASSWORD:
```
luckily there's probably no overlap between github farmers and API-probing bots

ludamad commented on yarn-project/aztec/scripts/aztec.sh:
```diff
@@ -1,20 +1,27 @@
 case $cmd in
   test)
-    export LOG_LEVEL="${LOG_LEVEL:-info}"
+    export LOG_LEVEL="${LOG_LEVEL:-error}"
```
hmm not having LOG_LEVEL=info is clean but I wonder if it will make it harder when people report their issues

charlielye replied:
Unless I'm missing something, I thought the TXE should basically be operating transparently. It's obviously useful to see error messages, but INFO is basically debug in this context right? It's nargo that's giving all the useful feedback? Trivial to revert later if needed.

ludamad commented on aztec-up/bin/0.0.1/install:
```diff
@@ -0,0 +1,152 @@
+# This script is called by aztec-up to install a specific version of the Aztec toolchain.
+# It expects VERSION and INSTALL_URI to be set.
+set -euo pipefail
+shopt -s inherit_errexit
```
nit: At first guess I would say 0.0.1/install is not the current source

charlielye committed:
feat: aztec-up can be used to manage multiple aztec versions

`aztec-up` can be used to manage multiple versions.

```
$ aztec-up
aztec-up - Aztec version manager

Usage: aztec-up [command] [options]

Commands:
  install <version>         Install a version and switch to it
  use [<version>]           Switch to an installed version (or read from .aztecrc)
  list                      List installed versions
  uninstall <version>       Remove an installed version
  self-update               Update aztec-up itself to the latest version

Options:
  -h, --help                Show this help message

Examples:
  aztec-up install          Install the latest version
  aztec-up install nightly  Install the latest nightly version
  aztec-up install 0.85.0   Install a specific version
  aztec-up use 0.85.0       Switch to version 0.85.0
  aztec-up use              Read version from .aztecrc and switch to it
  aztec-up list             Show all installed versions
  aztec-up self-update      Update aztec-up to latest
```

If a `.aztecrc` file is present in the cwd, `aztec` will automatically switch to that version in that run context.

```
t % aztec --version
0.0.1-commit.1142ef1
t % cd fooproj
t % cat .aztecrc
0.0.1-commit.6d3c34e
t % aztec --version
0.0.1-commit.6d3c34e
t %
```

# Testing

You can install `aztec-up` with:
```
VERSION=0.0.1-commit.6d3c34e bash -i <(curl -s https://install.aztec.network/0.0.1-commit.6d3c34e/aztec-install)
```

At present you can't list available remote versions. The code does currently support listing available remote aliases, but they're not deployed yet. You can install another version with e.g.
```
aztec-up install 0.0.1-commit.1142ef1
```

Then list versions with:

```
t % aztec-up list
Installed versions:

* 0.0.1-commit.1142ef1 (current)
  0.0.1-commit.6d3c34e

Aliases:
  (could not fetch aliases)
```
